File: leet/stream_of_characters/stream_of_characters.js
#!/usr/bin/env node

var TreeNode = function() {
    this.children = {};
    this.value = false;
};


/**
 * @param {string[]} words
 */
var StreamChecker = function(words) {
    this.root = new TreeNode();
    this.stream = [];
    for (let word of words)
    { 
        word = word.split('').reverse().join('');
        let node = this.root;
        for (let char of word)
        {
            if (!node.children[char])
                node.children[char] = new TreeNode();
            node = node.children[char];
        }
        node.value = true;
    }
};

/** 
 * @param {character} letter
 * @return {boolean}
 */
StreamChecker.prototype.query = function(letter) {
    this.stream.unshift(letter);
    let node = this.root;
    for (let char of this.stream)
    {
        if (node.children[char])
        {
            node = node.children[char];
            if (node.value)
                return true;
        }
        else
            break;
    }
    return false;
};

/** 
 * Your StreamChecker object will be instantiated and called as such:
 * var obj = new StreamChecker(words)
 * var param_1 = obj.query(letter)
 */

words = ["cd","f","kl"];
stream = [["a"],["b"],["c"],["d"],["e"]]
obj = new StreamChecker(words)
for (char of stream)
{
    console.log(char[0], obj.query(char[0]));
}

File: leet/stream_of_characters/stream_of_characters.java
class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isWord = false;
    
    public void insert(String word)
    {
        TrieNode node = this;
        for (char letter : word.toCharArray())
        {
           if (node.children.get(letter) == null)
               node.children.put(letter, new TrieNode());
            node = node.children.get(letter);
        }
        node.isWord = true;
    }
}

class StreamChecker {

    TrieNode root = new TrieNode();
    StringBuilder stream = new StringBuilder();
    
    public StreamChecker(String[] words) {
        for (String word : words)
            this.root.insert(new StringBuilder().append(word).reverse().toString());
    }
    
    public boolean query(char letter) {
        TrieNode node = this.root;
        this.stream.insert(0, letter);
        for (int i = 0; i < this.stream.length(); i++)
        {
            char c = this.stream.charAt(i);
            if((node = node.children.getOrDefault(c, null)) == null)
                break;
            if (node.isWord)
                return true;
        }
        return false;
    }
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker obj = new StreamChecker(words);
 * boolean param_1 = obj.query(letter);
 */
